generate/schemagen/gen/groups.go

```diff
@@ -0,0 +1,143 @@
+type FacadeGroup string
+
+const (
+	// Latest gets the latest facades from all the facades.
+	Latest FacadeGroup = "latest"
+	// All gets all the facades no matter the version.
+	All FacadeGroup = "all"
+	// Client facades returns just the client facades along with some required
+	// facades that the client can use.
+	Client FacadeGroup = "client"
+	// JIMM facade group defines a very select set of facades that only work
+	// with JIMM. This does not include the JIMM facade as defined in JIMM.
+	JIMM FacadeGroup = "jimm"
+)
+
+// ParseFacadeGroup will attempt to parse the facade group
+func ParseFacadeGroup(s string) (FacadeGroup, error) {
+	switch s {
+	case "latest", "all", "client", "jimm":
+		return FacadeGroup(s), nil
+	default:
+		return FacadeGroup(""), errors.NotValidf("facade group")
+	}
+}
+
+// Filter the facades based on the group.
+func Filter(g FacadeGroup, facades []facade.Details, registry Registry) []facade.Details {
+	switch g {
+	case Latest:
+		return latestFacades(facades)
+	case All:
+		return allFacades(facades)
+	case Client:
+		return clientFacades(facades, registry)
+	case JIMM:
+		return jimmFacades(facades)
+	}
+	return facades
+}
+
+func latestFacades(facades []facade.Details) []facade.Details {
+	latest := make(map[string]facade.Details)
+	for _, facade := range facades {
+		if f, ok := latest[facade.Name]; ok && facade.Version < f.Version {
+			continue
+		}
+		latest[facade.Name] = facade
+	}
+	latestFacades := make([]facade.Details, 0, len(latest))
+	for _, v := range latest {
+		latestFacades = append(latestFacades, v)
+	}
+	return latestFacades
+}
+
+func allFacades(facades []facade.Details) []facade.Details {
+	return facades
+}
+
+func clientFacades(facades []facade.Details, registry Registry) []facade.Details {
+	required := map[string]struct{}{
+		"Admin":               {},
+		"AllWatcher":          {},
+		"AllModelWatcher":     {},
+		"ModelSummaryManager": {},
+		"Pinger":              {},
+	}
+
+	results := make([]facade.Details, 0)
+	latest := latestFacades(facades)
+	for _, v := range latest {
+		if _, ok := required[v.Name]; ok {
+			results = append(results, v)
+			continue
+		}
+
+		var objType *rpcreflect.ObjType
+		kind, err := registry.GetType(v.Name, v.Version)
+		if err == nil {
+			objType = rpcreflect.ObjTypeOf(kind)
+		} else {
+			objType = rpcreflect.ObjTypeOf(v.Type)
+			if objType == nil {
+				continue
+			}
+		}
+		pkg := packageName(objType.GoType())
+		if !strings.HasPrefix(pkg, "github.com/juju/juju/apiserver/facades/client/") {
+			continue
+		}
+		results = append(results, v)
+	}
+	return results
+}
+
+func jimmFacades(facades []facade.Details) []facade.Details {
+	required := map[string][]int{
+		"Bundle":              {1},
+		"Cloud":               {1, 2, 3, 4, 5},
+		"Controller":          {3, 4, 5, 6, 7, 8, 9},
+		"ModelManager":        {2, 3, 4, 5},
+		"ModelSummaryManager": {1},
+		"Pinger":              {1},
+		"UserManager":         {1},
```

Comment:
Will these versions need to be updated when a facade version is rev'd?  Please add a comment if needed, or not needed.

Reply:
Yes, yes they will!